dimod/binary/binary_quadratic_model.py

```diff
@@ -702,105 +702,125 @@ def add_linear_equality_constraint(
-            with each tuple constituting a term in the summation.
-        """
-
-        if isinstance(terms, Iterator):
-            terms = list(terms)
-
-        if int(constant) != constant or int(lb) != lb or int(ub) != ub or any(
-                int(bias) != bias for _, bias in terms):
-            warnings.warn("For constraints with fractional coefficients, "
-                          "multiply both sides of the inequality by an "
-                          "appropriate factor of ten to attain or "
-                          "approximate integer coefficients. ")
-
-        terms_upper_bound = sum(v for _, v in terms if v > 0)
-        terms_lower_bound = sum(v for _, v in terms if v < 0)
-        ub_c = min(terms_upper_bound, ub - constant)
-        lb_c = max(terms_lower_bound, lb - constant)
-
-        if terms_upper_bound <= ub_c and terms_lower_bound >= lb_c:
-            warnings.warn(
-                f'Did not add constraint {label}.'
-                ' This constraint is feasible'
-                ' with any value for state variables.')
-            return []
-
-        if ub_c < lb_c:
-            raise ValueError(
-                f'The given constraint ({label}) is infeasible with any value'
-                ' for state variables.')
-
-        slack_upper_bound = int(ub_c - lb_c)
-        if slack_upper_bound == 0:
-            self.add_linear_equality_constraint(terms, lagrange_multiplier, -ub_c)
-            return []
-        else:
-            slack_terms = []
-            zero_constraint = False
-            if cross_zero:
-                if lb_c > 0 or ub_c < 0:
-                    if ub_c-slack_upper_bound > 0:
-                        zero_constraint = True
-
-            num_slack = int(np.floor(np.log2(slack_upper_bound)))
-            slack_coefficients = [2 ** j for j in range(num_slack)]
-            if slack_upper_bound - 2 ** num_slack >= 0:
-                slack_coefficients.append(slack_upper_bound - 2 ** num_slack + 1)
-
-            for j, s in enumerate(slack_coefficients):
-                sv = self.add_variable(f'slack_{label}_{j}')
-                slack_terms.append((sv, s))
-
-            if zero_constraint:
-                sv = self.add_variable(f'slack_{label}_{num_slack + 1}')
-                slack_terms.append((sv, ub_c - slack_upper_bound))
-
-        self.add_linear_equality_constraint(terms + slack_terms,
-                                            lagrange_multiplier, -ub_c)
-        return slack_terms
+                self, terms: Iterable[Tuple[Variable, int]],
+                lagrange_multiplier: Bias,
+                label: str,
+                constant: int = 0,
+                lb: int = np.iinfo(np.int64).min,
+                ub: int = 0,
+                cross_zero: bool = False,
+                penalization_method: str = "slack",
+        ) -> Iterable[Tuple[Variable, int]]:
+            """Add a linear inequality constraint as a quadratic objective.
```

Comment:
It looks like everything got indented by an additional tab?

Reply:
I fixed it! @arcondello